scripts/src/java/org/oppia/android/scripts/maven/GenerateLicenseTexts.kt

```diff
@@ -0,0 +1,423 @@
+  }
+
+  rootResourcesElement.appendChild(arrayElement)
+  doc.appendChild(rootResourcesElement)
+  doc.xmlStandalone = true
+  getTransformer().transform(DOMSource(doc), StreamResult(file))
+}
+
+private fun writeDependenciesLicenseNamesArray(
+  pathToResourceXml: String,
+  copyrightLicenseList: Set<CopyrightLicense>,
+  mavenDependenciesList: List<MavenDependency>
+) {
+  val file = File(pathToResourceXml)
+  val docBuilder: DocumentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder()
+  val doc: Document = docBuilder.newDocument()
+  val rootResourcesElement: Element = doc.createElement("resources")
+  val licenseMap = hashMapOf<String, String>()
+
+  // Add all license names.
+  for (index in copyrightLicenseList.indices) {
+    val licenseLink = copyrightLicenseList.elementAt(index).licenseLink
+    val licenseName = copyrightLicenseList.elementAt(index).licenseName
+    val stringElement = doc.createElement("string")
+    stringElement.setAttribute("name", "license_name_$index")
+    stringElement.appendChild(doc.createTextNode(licenseName))
+    rootResourcesElement.appendChild(stringElement)
+    licenseMap[licenseLink] = "license_name_$index"
+  }
+
+  // Add all string-arrays of license names for each dependency.
+  for (index in mavenDependenciesList.indices) {
+    val dependency = mavenDependenciesList[index]
+    val stringArrayElement = doc.createElement("string-array")
+    stringArrayElement.setAttribute("name", "third_party_dependency_licenses_names_$index")
+    dependency.licenseList.forEach { license ->
+      val licenseLink = if (license.primaryLinkType == PrimaryLinkType.SCRAPE_DIRECTLY ||
+        license.primaryLinkType == PrimaryLinkType.SHOW_LINK_ONLY
+      ) {
+        license.primaryLink
+      } else {
+        license.alternativeLink
+      }
+      val indexOfLicenseName = licenseMap[licenseLink]
+      val stringItemElement = doc.createElement("item")
+      stringItemElement.appendChild(doc.createTextNode("@string/$indexOfLicenseName"))
+      stringArrayElement.appendChild(stringItemElement)
+    }
+    rootResourcesElement.appendChild(stringArrayElement)
+  }
+
+  // Add an array that contains string-arrays of license names for each dependnecy.
+  val arrayElement = doc.createElement("array")
+  arrayElement.setAttribute("name", "third_party_dependency_license_names_array")
+  for (index in mavenDependenciesList.indices) {
+    val itemElement = doc.createElement("item")
+    itemElement.appendChild(
+      doc.createTextNode("@array/third_party_dependency_licenses_names_$index")
+    )
+    arrayElement.appendChild(itemElement)
+  }
+
+  rootResourcesElement.appendChild(arrayElement)
+  doc.appendChild(rootResourcesElement)
+  doc.xmlStandalone = true
+  getTransformer().transform(DOMSource(doc), StreamResult(file))
+}
+
+fun retrieveAllLicensesSet(
+  mavenDependencyList: List<MavenDependency>
+): Set<CopyrightLicense> {
+  val copyrightLicensesSet = mutableSetOf<CopyrightLicense>()
```

Comment:
This is another case we can avoid needing to use a mutable set.

Reply:
Sorry, I got a bit confused here, can you give an example here?

Reply:
This was addressed in a later comment (the idea is to use methods like ``map{}`` to avoid needing to use mutable structures).